source/Test.cpp

```diff
@@ -192,12 +274,81 @@ void Test::Step(Context &context, UI &menuPanels, UI &gamePanels, PlayerInfo &pl
 		return;
 	}
 	
-	const TestStep &stepToRun = steps[context.stepToRun];
+	// All processing was done just before this step started.
+	context.unprocessedInput = false;
+	context.branchesSinceGameStep.clear();
 	
-	// Exit with error on a failing testStep.
-	const string &stepTypeName = STEPTYPE_TO_TEXT.at(stepToRun.stepType);
-	
-	Fail("Test step " + to_string(context.stepToRun) + " (" + stepTypeName + ") failed");
+	bool continueGameLoop = false;
+	while(context.stepToRun < steps.size() && !continueGameLoop)
+	{
+		// Fail if we encounter a watchdog timeout
+		if(context.watchdog == 1)
+			Fail(context, player, "watchdog timeout");
+		else if(context.watchdog > 1)
+			--(context.watchdog);
+		
+		const TestStep &stepToRun = steps[context.stepToRun];
+		switch(stepToRun.stepType)
+		{
+			case TestStep::Type::APPLY:
+				stepToRun.conditions.Apply(player.Conditions());
+				++(context.stepToRun);
+				break;
+			case TestStep::Type::ASSERT:
+				if(!stepToRun.conditions.Test(player.Conditions()))
+					Fail(context, player, "asserted false");
+				++(context.stepToRun);
+				break;
+			case TestStep::Type::BRANCH:
+				// If we encounter a branch entry twice, then resume the gameloop before the second encounter.
+				if(context.branchesSinceGameStep.count(context.stepToRun))
+				{
```

Comment:
Whats the reasoning for this?

Reply:
This deals with loops that wait for some condition to become true. If we don't regularly return to the gameloop from such "wait loops", then the game would never have the chance to proceed with actions that can make the condition true.

Reply:
I added this explanation as a comment in the code.